src/shared/utilities/timeoutUtils.ts

```diff
@@ -118,3 +123,43 @@ export async function waitUntil<T>(
         await new Promise(r => setTimeout(r, opt.interval))
     }
 }
+
+/**
+ * Utility function to wrap a Timeout token around a Promise.
+ *
+ * @param promise The promise to use a Timeout with
+ * @param timeout A Timeout token that will race against the promise
+ * @param opt.gracefulTermination Allows the promise to be resolved undefined (default: true)
+ * @param opt.onExpire Callback for when the promise expired. The callback can return a value
+ * @param opt.onCancel Callback for when the promise was cancelled. The callback can return a value
+ *
+ * @returns A Promise that resolves into a valid return value, or rejects when the Timeout was cancelled or expired.
+ */
+export function createTimedPromise<T>(
+    promise: Promise<T>,
+    timeout: Timeout,
+    opt: { gracefulTermination?: boolean; onExpire?: () => T | undefined; onCancel?: () => T | undefined } = {}
+): Promise<T | undefined> {
+    opt.gracefulTermination = opt.gracefulTermination ?? true
```

Comment:
nit: could swap this to `forcefulTermination` or something and just check later `if (opt.forcefulTermination)`